core/src/main/java/jenkins/tasks/SimpleBuildStep.java

```diff
@@ -93,28 +93,76 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
 
     /**
      * Run this step.
-     * @param run a build this is running as a part of
+     * <p>
+     * This method <strong>must</strong> be overridden when this step requires a workspace context. If such a context is
+     * <em>not</em> required, it does not need to be overridden; it will then forward to
+     * {@link #perform(Run, EnvVars, TaskListener)}.
+     *
+     * @param run       a build this is running as a part of
      * @param workspace a workspace to use for any file operations
-     * @param env environment variables applicable to this step
-     * @param launcher a way to start processes
-     * @param listener a place to send output
+     * @param env       environment variables applicable to this step
+     * @param launcher  a way to start processes
+     * @param listener  a place to send output
+     * @throws AbstractMethodError  if this step requires a workspace context and this method is not overridden
      * @throws InterruptedException if the step is interrupted
-     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @throws IOException          if something goes wrong; use {@link AbortException} for a polite error
```

Comment:
I do not think we generally try to align Javadoc.

```suggestion
     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
```

Reply:
It's something IntelliJ does; I can remove the indents manually if needed.